enhancements/builds/volume-mounted-resources.md

```diff
@@ -0,0 +1,374 @@
+
+Other options like `SubPath` and `MountPropagation` may be considered for support in the future.
+
+When the build pod is constructed, all *input volumes* will be mounted in a subdirectory of a fixed location in the build pod's main container - `/var/run/openshift.io/volumes`.
+The OpenShift builder process will infer the destination directory of these volumes inside the *buildah runtime environment* from the `Build` API object (already available via an environment variable, serialized as JSON).
+The logic that invokes buildah will then pass the buildah volume mounts as transient bind mount arguments.
+
+#### API Example
+
+```go
+// SourceBuildStrategy defines input parameters specific to an Source build.
+type SourceBuildStrategy struct {
+  
+  // existing API
+  ...
+  // Note that these fields will also be added to the DockerBuildStrategy API
+
+  // Volumes is a list of input volumes that can be mounted into buildah's runtime environment.
+  // Only a subset of Kubernetes Volume sources are supported by builds.
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes
+  Volumes []BuildVolume
+
+  // VolumeMounts are volumes to mount into buildah's runtime environment at the specified mount path.
+  VolumeMounts []BuildVolumeMount
+}
+```
+
+```go
+// BuildVolume describes a volume that is made available to build pods, such that it can be mounted into buildah's runtime environment.
+// Only a subset of Kubernetes Volume sources are supported by builds.
+type BuildVolume struct {
+
+  // Volume's name.
+	// Must be a DNS_LABEL, unique within the pod, and cannot collide with volumes that are always added by the build controller.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string 
+
+	// BuildVolumeSource represents the location and type of the mounted volume.
+	BuildVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+}
+
+type BuildVolumeSourceType string
+
+const (
+  BuildVolumeSourceTypeSecret = "Secret"
+  BuildVolumeSourceTypeConfigMap = "ConfigMap"
+)
+
+// Represents the source of a volume to mount.
+// Only one of its members may be specified.
+// This is a subset of the core Kubernetes VolumeSource.
+type BuildVolumeSource struct {
+
+  // Type is the type for the volume source.
+  // Type must match the populated volume source, and if not specified will be inferred on admission.
+  // Only one type of volume source can be specified.
+  Type BuildVolumeSourceType
+
+  // Secret represents a secret that should populate this volume.
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+  // +optional
+  Secret *kapi.SecretVolumeSource
+
+  // ConfigMap represents a configMap that should populate this volume
+  // +optional
+  ConfigMap *kapi.ConfigMapVolumeSource
+}
+```
+
+```go
+// BuildVolumeMount describes a mounting of a Volume within buildah's runtime environment.
+type BuildVolumeMount struct {
```

Comment:
/cc @nalind 

For mount options supported by buildah.

Reply:
The library accepts options for controlling propagation ("shared", "rshared", "private", "rprivate", "slave", "rslave"), regular options ("rw, "ro", "dev"/"nodev", "suid"/"nosuid", "exec"/"noexec"), and SELinux labeling controls that I think end up being ignored since we're already in a container ("Z", "z").
There are also a couple of custom ones: "U" to have the ownership of the volume's contents changed to match the current USER value, and "O" which causes the volume source to be used as the lower layer for an overlay mount at the mount point, using a per-run temporary directory as the upper (write) layer.  I don't know that either of those is particularly useful in builds, and of course their functionality willl be limited in the future by the permissions that the builder binary has.

Reply:
Good to know - we will start with ReadOnly `ro` as the only supported option.